path_role.go

```diff
@@ -243,39 +273,99 @@ func (b *azureAuthBackend) pathRoleCreateUpdate(ctx context.Context, req *logica
+		if !ok {
+			policiesRaw, ok = data.GetOk("policies")
+			if ok {
+				role.Policies = policyutil.ParsePolicies(policiesRaw)
+				role.TokenPolicies = role.Policies
+			}
+		} else {
+			_, ok = data.GetOk("policies")
+			if ok {
+				role.Policies = role.TokenPolicies
+			} else {
+				role.Policies = nil
+			}
+		}
 
-	periodRaw, ok := data.GetOk("period")
-	if ok {
-		role.Period = time.Duration(periodRaw.(int)) * time.Second
-	} else if req.Operation == logical.CreateOperation {
-		role.Period = time.Duration(data.Get("period").(int)) * time.Second
-	}
-	if role.Period > b.System().MaxLeaseTTL() {
-		return logical.ErrorResponse(fmt.Sprintf("'period' of '%q' is greater than the backend's maximum lease TTL of '%q'", role.Period.String(), b.System().MaxLeaseTTL().String())), nil
-	}
+		numUsesRaw, ok := data.GetOk("token_num_uses")
+		if !ok {
+			numUsesRaw, ok = data.GetOk("num_uses")
+			if ok {
+				role.NumUses = numUsesRaw.(int)
+				role.TokenNumUses = role.NumUses
+			}
+		} else {
+			_, ok = data.GetOk("num_uses")
+			if ok {
+				role.NumUses = role.TokenNumUses
+			} else {
+				role.NumUses = 0
+			}
+		}
 
-	if tokenNumUsesRaw, ok := data.GetOk("num_uses"); ok {
-		role.NumUses = tokenNumUsesRaw.(int)
-	} else if req.Operation == logical.CreateOperation {
-		role.NumUses = data.Get("num_uses").(int)
-	}
-	if role.NumUses < 0 {
-		return logical.ErrorResponse("num_uses cannot be negative"), nil
+		ttlRaw, ok := data.GetOk("token_ttl")
+		if !ok {
+			ttlRaw, ok = data.GetOk("ttl")
+			if ok {
+				role.TTL = time.Duration(ttlRaw.(int)) * time.Second
+				role.TokenTTL = role.TTL
+			}
+		} else {
+			_, ok = data.GetOk("ttl")
+			if ok {
+				role.TTL = role.TokenTTL
+			} else {
+				role.TTL = 0
+			}
+		}
+
+		maxTTLRaw, ok := data.GetOk("token_max_ttl")
+		if !ok {
+			maxTTLRaw, ok = data.GetOk("max_ttl")
+			if ok {
+				role.MaxTTL = time.Duration(maxTTLRaw.(int)) * time.Second
+				role.TokenMaxTTL = role.MaxTTL
+			}
+		} else {
+			_, ok = data.GetOk("max_ttl")
+			if ok {
+				role.MaxTTL = role.TokenMaxTTL
+			} else {
+				role.MaxTTL = 0
+			}
+		}
+
+		periodRaw, ok := data.GetOk("token_period")
+		if !ok {
+			periodRaw, ok = data.GetOk("period")
+			if ok {
+				role.Period = time.Duration(periodRaw.(int)) * time.Second
+				role.TokenPeriod = role.Period
+			}
+		} else {
+			_, ok = data.GetOk("period")
+			if ok {
+				role.Period = role.TokenPeriod
+			} else {
+				role.Period = 0
+			}
+		}
 	}
 
-	if tokenTTLRaw, ok := data.GetOk("ttl"); ok {
-		role.TTL = time.Duration(tokenTTLRaw.(int)) * time.Second
-	} else if req.Operation == logical.CreateOperation {
-		role.TTL = time.Duration(data.Get("ttl").(int)) * time.Second
+	if role.TokenPeriod > b.System().MaxLeaseTTL() {
+		return logical.ErrorResponse(fmt.Sprintf("token period of '%q' is greater than the backend's maximum lease TTL of '%q'", role.TokenPeriod.String(), b.System().MaxLeaseTTL().String())), nil
```

Comment:
```suggestion
		return logical.ErrorResponse("token period of %q is greater than the backend's maximum lease TTL of %q", role.TokenPeriod.String(), b.System().MaxLeaseTTL().String()), nil
```

Reply:
ErrorResponse takes in a bare string...

Reply:
It can also take a format string plus  parameters. And %q already adds wrapping quotes, so the extras were removed.

Reply:
...since when? :-)

Reply:
I don't really know why it had quotes plus %q and didn't really look, I just kept the error as is and just adjusted the parameter